pkg/pillar/cas/containerd.go

```diff
@@ -26,58 +46,221 @@ func (c *containerdCAS) CheckBlobExists(blobHash string) bool {
-		return -1, fmt.Errorf("GetBlobSize: Exception while getting size of blob: %s. %s", blobHash, err.Error())
+		return nil, fmt.Errorf("GetBlobInfo: Exception while getting size of blob: %s. %s", blobHash, err.Error())
 	}
-	return info.Size, nil
+
+	return &BlobInfo{
+		Digest: info.Digest.String(),
+		Size:   info.Size,
+		Labels: info.Labels,
+	}, nil
+
 }
 
-//ListBlobs: returns list of blobs where each entry is of format sha256:<hash>.
-func (c *containerdCAS) ListBlobs() ([]string, error) {
+//ListBlobInfo: returns list of BlobInfo for all the blob present in CAS
+func (c *containerdCAS) ListBlobInfo() ([]*BlobInfo, error) {
 	infos, err := containerd.CtrListBlobInfo()
 	if err != nil {
-		return nil, fmt.Errorf("ListBlobs: Exception while getting blob list. %s", err.Error())
+		return nil, fmt.Errorf("ListBlobInfo: Exception while getting blob list. %s", err.Error())
 	}
-	blobDigests := make([]string, 0)
+	blobInfos := make([]*BlobInfo, 0)
 	for _, info := range infos {
-		blobDigests = append(blobDigests, info.Digest.String())
+		blobInfos = append(blobInfos, &BlobInfo{
+			Digest: info.Digest.String(),
+			Size:   info.Size,
+			Labels: info.Labels,
+		})
 	}
-	return blobDigests, nil
+	return blobInfos, nil
 }
 
-//IngestBlob: reads the blob as raw data from arg 'reader' and validates the read blob's hash with the provided arg 'blobHash'.
-// If the validation succeeds, then the blob is ingested, else the given blob is rejected and an error is returned.
-//Arg 'blobHash' should be of format sha256:<hash>.
-//Arg 'ctx' must contain context with 'contextID':string with which we can relate two different blobs and
-// 'expires':time.Duration to set a validity duration for the context
-func (c *containerdCAS) IngestBlob(ctx context.Context, blobHash string, reader io.Reader) error {
-	leaseOpts := make([]leases.Opt, 0)
-	var leaseID string
-	if ctx.Value("contextID") == nil {
-		return fmt.Errorf("IngestBlob: context does not have 'contextID'")
+// IngestBlob: parses the given one or more `blobs` (BlobStatus) and for each blob reads the blob data from
+// BlobStatus.Path and ingests it into CAS's blob store.
+// Returns a list of loaded BlobStatus and an error is thrown if the read blob's hash does not match with the
+// respective BlobStatus.Sha256 or if there is an exception while reading the blob data.
+// In case of exception, the returned list of loaded blobs will contain all the blob that were loaded until that point.
+func (c *containerdCAS) IngestBlob(blobs ...*types.BlobStatus) ([]*types.BlobStatus, error) {
+	var (
+		index          *ocispec.Index
+		indexHash      string
+		manifests      = make([]*ocispec.Manifest, 0)
+		manifestHashes = make([]string, 0)
+	)
+	loadedBlobs := make([]*types.BlobStatus, 0)
+
+	//Step 1: Load blobs into CAS
+	for _, blob := range blobs {
+		//Process the blob only if its not loaded already
+		if blob.State == types.LOADED {
+			log.Infof("IngestBlob(%s): Blob as its already loaded", blob.Sha256)
```

Comment:
I think the message here is missing something. Was this supposed to be, "Not loading blob as its already loaded"?

Reply:
Changed.